Enable JuliaSyntax.jl as the default Julia parser

c42f opened the pull request:
This enables the use of JuliaSyntax.jl as the Julia parser for the runtime, which greatly improves parser error messages in various cases. ~~As an experimental feature, this is only enabled if the `JULIA_USE_NEW_PARSER=true` environment variable is set.~~ Update: it's the default!

![20220817_17h14m06s_grim](https://user-images.githubusercontent.com/601473/185057712-31deb848-e545-4f03-8fe7-fbb2462f9ae4.png)

For now I've installed it after bootstrapping because this is the simplest (flisp parser still used for bootstrap). In the future we can solve bootstrapping but this isn't necessary right away.


## API Versioning: Stdlib woes vs vendoring

The simplest way to get this working immediately was as a stdlib. However in retrospect this has several downsides:
* As only version JuliaSyntax-0.1, the `JuliaSyntax` API is not at all stable. The only stable API for using this from ` Base` should be `Meta.parse()` and `Expr`. Therefore, I think users should not really be able to do `using JuliaSyntax` and get the version bundled with Base.
* For use in tooling like the language server, it's necessary for the JuliaSyntax API (custom syntax tree types etc) to be able to evolve with its own versioning scheme independent from whatever version of `Base` someone is using. Users should be able to install a "normal" version of `JuliaSyntax` via `Pkg`.

So I think having this as a stdlib doesn't really make sense and we are left with some kind of vendoring approach. I think what I'll try next is to include the parser at the end of building `Base`.

Having said all that, it should be easy to try this out already :-)

Fixes #31449

inkydragon commented on deps/checksums/JuliaSyntax-25b737b1bbce259451f6d79997a25c541e7ab8c1.tar.gz/sha512:
```diff
@@ -0,0 +1 @@
+13af8fb86e9286ce95c47811f5d37f18013a67ce4a3b1b8fdc97d283a8898f14584db2dc441519f4c645e20d622b8ffb9613f2fc141e4622b26f362f17a8c2f7
```
CI error: https://buildkite.com/julialang/julia-master/builds/14799#0182ab15-7f58-4b9e-b57a-56addf3bed4d/552-2397

```
  ERROR: sha512 checksum failure on JuliaSyntax-25b737b1bbce259451f6d79997a25c541e7ab8c1.tar.gz, should be:
      13af8fb86e9286ce95c47811f5d37f18013a67ce4a3b1b8fdc97d283a8898f14
      584db2dc441519f4c645e20d622b8ffb9613f2fc141e4622b26f362f17a8c2f7
  But `sha512sum /cache/build/default-amdci5-1/julialang/julia-master/stdlib/srccache/JuliaSyntax-25b737b1bbce259451f6d79997a25c541e7ab8c1.tar.gz | awk '{ print $1; }'` results in:
      762532fcf6ec954165095e1892901d8afeff7b11fb4997aac2adbe0a7d572180
      d0eb49d1fa66edbf2559154206726cedd3c9db3362415e29836e1031897c6bcc
  This can happen due to bad downloads or network proxies, please check your
  network proxy/firewall settings and delete
  /cache/build/default-amdci5-1/julialang/julia-master/stdlib/srccache/JuliaSyntax-25b737b1bbce259451f6d79997a25c541e7ab8c1.tar.gz
  to force a redownload when you are ready
```

local test:

```sh
woclass@wos-PC:/mnt/v/tmp$ curl https://api.github.com/repos/JuliaLang/JuliaSyntax.jl/tarball/25b737b1bbce259451f6d79997a25c541e7ab8c1 -o JuliaSyntax.jl-25b737b1bbce259451f6d79997a25c541e7ab8c1.tar.gz
woclass@wos-PC:/mnt/v/tmp$ ls
JuliaSyntax.jl-25b737b1bbce259451f6d79997a25c541e7ab8c1.tar.gz  fasta.jl  fasta.txt  knucleotide.jl
woclass@wos-PC:/mnt/v/tmp$ sha512sum JuliaSyntax.jl-25b737b1bbce259451f6d79997a25c541e7ab8c1.tar.gz
cf83e1357eefb8bdf1542850d66d8007d620e4050b5715dc83f4a921d36ce9ce47d0d13c5d85f2b0ff8318d2877eec2f63b931bd47417a81a538327af927da3e  JuliaSyntax.jl-25b737b1bbce259451f6d79997a25c541e7ab8c1.tar.gz
```

JeffBezanson replied:
I get the checksum failure as well.

c42f replied:
The checksum `cf83e1357eefb8...` is just the checksum for the empty file, so something is weird with that `curl` command. It seems curl wasn't following the redirect for me, but adding `-L` didn't work.

I was able to download using `wget`:

```
wget https://api.github.com/repos/JuliaLang/JuliaSyntax.jl/tarball/25b737b1bbce259451f6d79997a25c541e7ab8c1
```

But then I was getting a different checksum each time, even though the archive content is the same.  Now when I run that exact same wget command nothing happens at all. So something *really* weird is going on here on the server side.

It's possible the weird behavior is due to me using the git tree hash rather than a commit hash in `JuliaSyntax.version`. Let me check/fix that.

c42f replied:
`cf83e1357eefb8bdf1542850d66d8007d620e4050b5715dc83f4a921d36ce9ce47d0d13c5d85f2b0ff8318d2877eec2f63b931bd47417a81a538327af927da3e` is the sha512 sum for an empty file :)

Anyway, so it seems this problem was due to using the tree hash rather than commit hash in `JuliaSyntax.version`.

Having done that, it seems github will do all sorts of weird stuff on the serverside, including
* Returning different `tgz` files each time but which unpack to the same tree
* Seemingly, randomly refusing to return a file at all and giving a 404 status instead

c42f replied:
That was confusing to debug.  Should be fixed now I hope.

inkydragon replied:
Ok, it works for me.

```sh
woclass@wos-PC:~$ wget -O JuliaSyntax-d3d8fdeaea37bcae05191cad50a326e608035d97.tar.gz  https://api.github.com/repos/JuliaLang/JuliaSyntax.jl/tarball/d3d8fdeaea37bcae05191cad50a326e608035d97
--2022-08-19 15:44:47--  https://api.github.com/repos/JuliaLang/JuliaSyntax.jl/tarball/d3d8fdeaea37bcae05191cad50a326e608035d97
Connecting to localhost:7890... connected.
Proxy request sent, awaiting response... 302 Found
Location: https://codeload.github.com/JuliaLang/JuliaSyntax.jl/legacy.tar.gz/d3d8fdeaea37bcae05191cad50a326e608035d97 [following]
--2022-08-19 15:44:47--  https://codeload.github.com/JuliaLang/JuliaSyntax.jl/legacy.tar.gz/d3d8fdeaea37bcae05191cad50a326e608035d97
Connecting to localhost:7890... connected.
Proxy request sent, awaiting response... 200 OK
Length: unspecified [application/x-gzip]
Saving to: ‘JuliaSyntax-d3d8fdeaea37bcae05191cad50a326e608035d97.tar.gz’

JuliaSyntax-d3d8fdeaea37bc     [  <=>                                  ] 122.64K   429KB/s    in 0.3s

2022-08-19 15:44:49 (429 KB/s) - ‘JuliaSyntax-d3d8fdeaea37bcae05191cad50a326e608035d97.tar.gz’ saved [125586]

woclass@wos-PC:~$ ll JuliaSyntax-d3d8fdeaea37bcae05191cad50a326e608035d97.tar.gz
-rw-r--r-- 1 woclass woclass 125586 Aug 19 15:44 JuliaSyntax-d3d8fdeaea37bcae05191cad50a326e608035d97.tar.gz
woclass@wos-PC:~$ md5sum JuliaSyntax-d3d8fdeaea37bcae05191cad50a326e608035d97.tar.gz
29ccd296e538403829ffd90b3cce67d8  JuliaSyntax-d3d8fdeaea37bcae05191cad50a326e608035d97.tar.gz
woclass@wos-PC:~$ sha512sum  JuliaSyntax-d3d8fdeaea37bcae05191cad50a326e608035d97.tar.gz
b3c26d6f7880f069b1f6f21ee3978cd3e797acd9ca6fb1e7ece28cf4dea576872bed3e8c3121d1d6cfe559518a17004595139f9978fd2591669cbe8128f7c1ac  JuliaSyntax-d3d8fdeaea37bcae05191cad50a326e608035d97.tar.gz
```

aviatesk commented on base/boot.jl:
```diff
@@ -824,9 +824,13 @@ Integer(x::Union{Float16, Float32, Float64}) = Int(x)
+# The internal jl_parse will call into Core._parse if not `nothing`.
 _parse = nothing
 
+function set_parser(parser)
+    global _parse = parser
+end
```
```suggestion
_setparser!(parser) = setglobal!(@__MODULE__, :_parse, parser)
```
is a bit simpler (and also make it more explicit that this is very internal function)?

c42f replied:
oh cool `setglobal!` is new to me.  Seems good.

vtjnash commented:
Can we drop the implied `Mmap` usage and dependency? Doing mmap leads to segfaults and bus errors, where a read would have just gotten a permission error or file truncation error, and can lead to other problems like that, with cleanup and continued edits by the user on the original file.

vtjnash commented:
It is really cool to see how small this is! And exciting!

c42f commented:
> Can we drop the implied Mmap usage and dependency

Yes :-) Already done on a WIP branch of mine, will be merged soon.

BTW, for anyone wondering what's happening here - all the development is happening over at JuliaSyntax.jl which is why this PR is just sitting around at the moment. I'll get back to this PR once I've sorted out a few remaining large usability problems (Particularly generating `Expr(:incomplete)` for REPL completions. Which isn't hard in itself, but has proven to be a bit of a rabbit hole leading to other things.)

oscardssmith commented:
Now that version 0.2 of JuliaSyntax has been released, can this PR be updated to use it? We have 2 weeks left if we want to merge this in the 1.9 timeframe.

PallHaraldsson commented:
>only enabled if the JULIA_USE_NEW_PARSER=true environment variable is set.

Can it be merged, and with it on by default? Would it be good for PkgEval, or should it alternatively be run with:
```
JULIA_USE_NEW_PARSER=true julia
```

I'm not necessarily thinking it should be in Julia 1.9-rc1, it just think it might be good to have it from now or the feature freeze (op to then) to check, and not have to explain the ENV to users, though this should also be a possibility:
```
JULIA_USE_NEW_PARSER=false julia
```

c42f commented:
> We have 2 weeks left if we want to merge this in the 1.9 timeframe.

Yep I've got julia-1.9 in mind for this. Will probably do a JuliaSyntax-0.3 release prior to that.

> Can it be merged, and with it on by default

I don't think we're ready to turn this on by default. `PkgEval` would be good though...

maleadt commented:
> `PkgEval` would be good though...

PkgEval currently doesn't configuring environment flags, so to test this either add a commit that enables the new parser by default, or add a Make flag that results in the new parser being used by default (which can then be set using the `configuration` keyword to Nanosoldier's `runtests`, resulting in a `Make.user` entry).

maleadt commented:
Tried running this under PkgEval (with latest JuliaSyntax.jl, not the one from this PR), but Julia fails to bootstrap:

```
error during bootstrap:
LoadError("sysimg.jl", 19, LoadError("/source/usr/share/julia/stdlib/v1.9/JuliaSyntax/src/JuliaSyntax.jl", 1, LoadError("/source/usr/share/julia/stdlib/v1.9/JuliaSyntax/src/parser_api.jl", 137, LoadError("/source/usr/share/julia/stdlib/v1.9/JuliaSyntax/src/parser_api.jl", 137, UndefRefError()))))
getproperty at ./Base.jl:37 [inlined]
getindex at ./refvalue.jl:56 [inlined]
docm at ./docs/Docs.jl:521
unknown function (ip: 0x7f90895bcc8d)
_jl_invoke at /source/src/gf.c:2447 [inlined]
ijl_apply_generic at /source/src/gf.c:2629
jl_apply at /source/src/julia.h:1866 [inlined]
do_apply at /source/src/builtins.c:730
@doc at ./boot.jl:534
```

I'll probably add an `environment` option to PkgEval then.

maleadt commented:
@nanosoldier `runtests(ALL, vs = "%self", configuration = (buildflags=["DEPS_GIT=JuliaSyntax"], environment=["JULIA_USE_NEW_PARSER=true"]))`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/46372#issuecomment-1268029710) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/ba2364a_vs_ba2364a/report.html).

maleadt commented:
Packages that seem to occur a lot in failure logs: Macrotools, Unitful, and CpuId.

oscardssmith commented:
`Unitful`, `AbstractAlgebra`, and `OrdinaryDiffeq` all have failures that I have filed issues about.

PallHaraldsson commented:
>In total, 8352 packages were tested, out of which 3596 succeeded, 3615 failed and 1141 were skipped.
>
>Testing took 5 hours, 3 minutes, 31 seconds (or, sequentially, 20 days, 6 hours, 50 minutes, 52 seconds to execute 16704 package tests suites).

This seems rather bad, but comparing to some random other PkgEval, about half failing isn't unusual (for all kinds of reasons including syntax):

https://github.com/JuliaLang/julia/pull/46372#issuecomment-1268029710

I'm amazed by how long it takes (and not scalable, with package ecosystem increasing), but is the timing maybe (only slightly?) better here?

maleadt commented:
> I'm amazed by how long it takes (and not scalable, with package ecosystem increasing), but is the timing maybe (only slightly?) better here?

The timings cannot be compared, as quite some packages now fail to precompile and thus never execute any test.

oscardssmith commented:
It is honestly kind of silly to use PkgEval to test this, since we could just test that everything parses the same between the 2 versions (which would allow us to skip loading the code and running the tests).

Krastanov commented:
The CpuId failure is due to something like commas after new lines:
```
julia> [1
       , 2]
ERROR: ParseError:
Error: Expected `]`
@ REPL[33]:2:1
[1
, 2]
```

oscardssmith commented:
yeah. this pr is still using 0.1  but should be using 0.2

oscardssmith commented:
fair :). I'm coming from the perspective that we'll probably have to run this between 10 and 100 times, at which point the balance might shift.

oscardssmith commented:
especially since the custom version could give more information per run since pkgeval doesn't give us errors for packages that have a failing dependency and does give us spurious errors for the thousand or so packages with bad CI

c42f commented:
> It is honestly kind of silly to use PkgEval to test this, since we could just test that everything parses the same between the 2 versions

We actually have some scripts for this in `JuliaSyntax/tools`. I need to revisit those and do another run them against all of General. Would also be good to hook them up to the syntax test case reduction tools I've got over there as well.  I think those scripts will be used for the bulk of the remaining compat work to come.

Regardless of those scripts, PkgEval will still be a useful tool because it tests the entire integration.

c42f commented:
## Status

I've updated this PR now to vendor JuliaSyntax into Base via the deps mechanism. This seemed cleanest because it's got to be downloaded before Base is built, so it's natural to have it unpacked as a dependency of the `julia-base` makefile target.

This resolves the major problems I saw with exposing this as a stdlib and I think we could now merge this PR right away so that it can get into Julia 1.9.

Obviously being experimental and still having known bugs, it will remain disabled by default for now.  A separate PR which removes the ENV check can be used to track progress toward full compatibility.

maleadt commented:
@oscardssmith Some packages also do run-time parsing of code, so its useful to actually run PkgEval for those. But I'm with @KristofferC here, this doesn't seem worth the development time. If we were short on compute I'd add a mode to PkgEval to only Pkg.precompile (and not actually run tests), but the machine is mostly idle (a single PkgEval run without `rr` only takes 5 hours), so even that doesn't seem worth it.

@c42f I'll let you trigger Nanosoldier when you know there's been fixes to the JuliaSyntax.jl repo 🙂

c42f commented:
> As noted above, that doesn't need to the new parser to be the default one anymore.

Perfect, thanks :-)

> I'll let you trigger Nanosoldier when you know there's been fixes to the JuliaSyntax.jl repo

Well the new parser is disabled by default so I don't think we need to block this PR on bugfixes or performance when the non-default JuliaSyntax config is enabled. But perhaps I'm missing something?

For this PR, I'd like someone to review the build system and vendoring I've set up here, then I think it's harmless enough to merge this. I'm not sure who to ping for this though! Our makefiles don't seem to have a definitive owner :sweat_smile:

c42f commented:
Oh btw, the test failures here are confusing but I guess unrelated. There's a profile stdlib test failure (i686-linux-gnu), some clipboard-related (windows) and filesystem-related failures (powerpc, aarch64).

maleadt commented:
> Well the new parser is disabled by default so I don't think we need to block this PR on bugfixes or performance when the non-default JuliaSyntax config is enabled. But perhaps I'm missing something?

Oh for sure, I just meant as a means to improve JuliaSyntax.jl

JeffBezanson commented:
I'd like to know what the system image size impact of this is. If it's significant, we might need an option to exclude the code from the build.

c42f commented:
@nanosoldier `runtests(ALL, vs = "%self", configuration = (buildflags=["DEPS_GIT=JuliaSyntax"], environment=["JULIA_USE_NEW_PARSER=true"]))`

nanosoldier commented:
[Your job](https://github.com/JuliaLang/julia/pull/46372#issuecomment-1280678234) failed. cc @maleadt

maleadt commented:
`make install` failed:

```
Oct 17 07:44:13 amdci8 bash[69705]:       From worker 2:        cp: cannot create regular file '/install/share/julia/base/JuliaSyntax/.git/objects/pack/pack-3676d1cb1e890803c88a8566efb0d7ae60442968.pack': Permission denied
Oct 17 07:44:13 amdci8 bash[69705]:       From worker 2:        cp: cannot create regular file '/install/share/julia/base/JuliaSyntax/.git/objects/pack/pack-3676d1cb1e890803c88a8566efb0d7ae60442968.idx': Permission denied
Oct 17 07:44:13 amdci8 bash[69705]:       From worker 2:        make: *** [Makefile:258: install] Error 1
```

c42f commented:
> system image size impact

Without JuliaSyntax in sysimage (commit 92e68c870794)
```
     20666212 corecompiler.ji
    111427187 sys.ji
    220552888 sys-o.a
    196755896 sys.so
```

With JuliaSyntax in sysimage (this PR):
```
     20659516 corecompiler.ji
    113788496 sys.ji
    238108968 sys-o.a
    212239016 sys.so
```

So sys.so grows by almost 8% or 15 MB which seems ... well this seems like an awful lot and I'm puzzled why it's quite this much.

aviatesk commented:
> So sys.so grows by almost 8% or 15 MB which seems ... well this seems like an awful lot and I'm puzzled why it's quite this much.

To me these numbers seem reasonable, as we are now precompiling lots of JuliaSyntax.jl code into sysimg, i.e. the # of precompiled signatures increases from `1723` on master to `2193` on this PR (rebased locally).
I think there are several ways to decrease this number (which will likely to decrease the sysimg size also), e.g. 1.) not to run the whole part of the JuliaSyntax test suite but only parts that covers common cases, 2.) improve type stabilities of JuliaSyntax (this is just a guess, but JuliaSyntax.jl may need to add bunch of `@nospecialize` annotations to suppress dynamic dispatches leading to excessive code generation).

PallHaraldsson commented:
So it has 2193-1723 = 470 "precompiled signatures"? Why would you run test suite to generate them? Can it have just the bare minimum or even none? Since it's not yet the default and you need to opt into using it, it's just code sitting there not useful for most people. It would though be nice to have it on by default (or when you enable it) to be fast, but then I guess none of the error paths need to be precompiled, if that's (easily) possible. I just know it has better error messages, likely why it's large.

I see in a precompile file:
```
    JuliaSyntax.enable_in_core!()
    Meta.parse("x+y+z-w .+ [a b c]")
```

oscardssmith commented:
we could merge this, but it isn't that useful yet as a relatively large number of packages will fail to load do to parsing issues in the dependency chain.

KristofferC commented:
> Can this be merged?

See https://github.com/JuliaLang/julia/pull/46372#issuecomment-1275493465

PallHaraldsson commented:
I suppose you mean the 8% growth (470 "precompiled signatures") is too large, but it's not clear it's too large 8for 1.9-DEV) or if Jeff feels so. It would only fail to load "relatively large number of packages" if enabled with the ENV (or on by default, which seems not considered), none will fail with it disabled(?). But a lot will be ok.

c42f commented:
Note this PR shouldn't be merged until https://github.com/JuliaLang/julia/pull/46372#issuecomment-1280754426 is resolved.

For sysimage size, most appears to be related to the data structures stored by Julia, not the native code, as reported by `size -A -d` and comparing the `.data` and `.text` sections:

|                    |section |  size       | size (MiB) |
|--------------------|--------|-------------|----------- |
|With JuliaSyntax    | native code (.text)   |  10529720   |  10.53 |
|                    | julia IR, etc (.data)   |  175972226  | 175.97 |
|Without JuliaSyntax | native code (.text   |  9098872    |   9.1 |
|                    | julia IR, etc  (.data)   |  164312325  | 164.31 |

The .text section shows the native code is only costing an additional 1.4 MiB here which makes much more sense to me. IIUC the .data section being a combination of IR, method tables, types, inference cache etc isn't something we can really control until we get separate compilation and can strip these out.

c42f commented:
@PallHaraldsson if you are keen to test this you can test it today with a custom sysimage and the latest (more recent and bugfixed) version of JuliaSyntax.jl. See https://github.com/JuliaLang/JuliaSyntax.jl#using-juliasyntax-as-the-default-parser for simple instructions. This will be a better experience for reporting bugs and getting them fixed more quickly, rather than having to wait for a new Base release.

This PR missed 1.9 due to needing a bunch more testing against General and bugfixing (also fixes to the build system still required here). But people can still try it out easily. Try the custom sysimage!

oscardssmith commented:
technically it hasn't missed 1.9 yet. we've been delaying the release for native code caching

PallHaraldsson commented:
>IIUC the .data section being a combination of IR, method tables, types, inference cache etc isn't something we can really control until we get separate compilation and can strip these out.

Are these, new in 1.8, relevant here:
```
* New option --strip-metadata to remove docstrings, source location information, and local variable names when building a system image (https://github.com/JuliaLang/julia/issues/42513).
* New option --strip-ir to remove the compiler's IR (intermediate representation) of source code when building a system image. The resulting image will only work if --compile=all is used, or if all needed code is precompiled (https://github.com/JuliaLang/julia/issues/42925).
```

I would be ok with merging in 1.9-DEV (off by default), as is, even with the sysimage larger (or with above). Yes, I can test that way:
>if you are keen to test this you can test it today with a custom sysimage

would it be identical; as far as startup (and parsing) speed? And yes I'm keen. A merge can always be reverted later...

It passes tests, Win64 is full of false alarms (so ok to merge and/or restart its tests?), basically all it is "No space left on device". And FreeBSD is too (plus tier 3 platform): "EOFError: read end of file".

ViralBShah commented:
We have a few more days to go for the 1.9 feature freeze, it appears. So, if this is close enough to get in (behind a feature flag), it would be great. I suppose it can release gated behind a feature flag for 1.9, but become the default in 1.10.

PallHaraldsson commented:
It's already behind a "feature flag", i.e. ENV controls, so just "merge me"?

c42f commented:
I chatted to @JeffBezanson yesterday about the status here and when to push through with merging this. His opinion, IIUC, was to delay merging until more JuliaSyntax bugs are sorted out upstream. Primarily

* having error messages for literal processing (eg, reporting invalid character literals, string escaping errors, etc)
* further improvements to generating `Expr(:incomplete)` for REPL usability
* perhaps also parsing all of General correctly? (There's currently ~20 known failures across the ~130000 .jl files in General, but there's probably others which the test scripts don't catch. Fixing all the little issues is a loong tail of obscure bugs.)

Hopefully I understood and listed Jeff's concerns correctly above.

ViralBShah commented:
I was just talking with @vtjnash about roadmap for 1.10 and 1.11, and one thing I was thinking out aloud was "It would be real nice if we can have the JuliaSyntax.jl work land in 1.10".  Is that possible?

oscardssmith commented:
See https://github.com/JuliaLang/JuliaSyntax.jl/issues/134. There are 7 remaining issues.

PallHaraldsson commented:
I think we should merge in 1.10 for sure, since it's opt-in with ENV anyway (real question if it should default to on in 1.10, should then be backported to 1.9, off by default there).

The current Julia parser fails in 3 cases and JuliaSyntax doesn't (my reading), yes, 7 remaining issues (some obscure?), would people be more likely to help and find possibly more bugs if merged to master?

c42f commented:
I think we can land it "real soon now" (TM). It just depends on making a call about how much compatibility is good enough.

To be honest the expected level of effort here has gone up a lot as I've tested more and more strictly. Fixing the long tail of obscure bugs has been an adventure :sweat_smile:  Hyrum's law very much applies.

The current status is
* In parsing my snapshot of General: out of 128900 files there are 35 `Expr` mismatches; 0.2% failure rate.
* In running the tests in `test/syntax.jl`, there's still failures due to expecting exact error message matches with the reference parser. This is unrealistic/undesired, but it would be nice to fix these and make all tests pass with the new parser enabled. Iv'e been working on this.
* `Expr(:incomplete)` generation could still use some work.

gbaraldi commented:
I think we should merge this  soon :tm:  and start tracking the failures in Base as well, so they get a bit more effort. I don't doubt we are going to find some issues but they might need widespread use to appear.

IanButterworth commented on base/Base.jl:
```diff
@@ -529,6 +533,9 @@ function __init__()
     if haskey(ENV, "JULIA_MAX_NUM_PRECOMPILE_FILES")
         MAX_NUM_PRECOMPILE_FILES[] = parse(Int, ENV["JULIA_MAX_NUM_PRECOMPILE_FILES"])
     end
+    if get(ENV, "JULIA_USE_NEW_PARSER", "false") == "true"
```
```suggestion
    if get_bool_env("JULIA_USE_NEW_PARSER", false) === true
```

o314 replied:
`get_bool_env` seems quite new . 
This is in our codebase

```julia
parse_from_env(::Type{Union{String,Nothing}}, name, default=""; is_src_nothing=isempty) =
    let val = get(ENV, name, default); is_src_nothing(val) ? nothing : val end
parse_from_env(::Type{String}, name, default="") = get(ENV, name, default)
parse_from_env(::Type{Vector{String}}, name, default="", dlm=",") =
    let v=get(ENV,name,default); split(v, dlm; keepempty=false) end
parse_from_env(::Type{Bool}, name, default="") = let v=get(ENV,name,default); lowercase(v) in ("1", "true") end
parse_from_env(::Type{Int}, name, default="") = let v=get(ENV,name,default); parse(Int, v) end
parse_from_env(::Type{URI}, name, default="") =
    let val = get(ENV, name, default) ;
        isempty(val) ?
            nothing :
            val |>                                                              # use norm_url ? # use URI
                val -> (startswith(val, r"\w+://") ? val : "https://$val") |>   # with scheme; alleviate RFC1808 from ENV
                val -> rstrip(val, '/') |> # without trailing slash                # <- URIs.normpath
                URI
    end
```

It better maps to parse and is more complete. Meanwhile your truthy/falsy map may be larger.

timholy commented:
Can you clarify whether any of the parse failures result in failure of the package to precompile or run correctly? Or is this a case where things mostly still work? If the packages are broken, my experience from the pkgimages work suggests it's better to fix known issues before merging, unless you know that all of them can be addressed by minor tweaks to the packages themselves.

https://github.com/JuliaDebug/Cthulhu.jl/pull/345 is excited to hear that this is close. Really grateful that you've tackled this.

IanButterworth reviewed (commented):
Seems like the new stuff added to the precompile step is outputting in a way that leaks. Perhaps to stderr when it's monitoring stdout, or something
```
Collecting and executing precompile statements
--
  | └ Collect (Basic: ◐ , REPL 10/40: ◐ 951) => Execute ◐ 68Test Summary:     \| Pass  Broken  Total   Time
  | Inline test cases \|  699      10    709  23.9s
  | Test Summary:     \| Pass  Total  Time
  | Trivia attachment \|    1      1  0.6s
  | Test Summary:                            \| Pass  Total  Time
  | Unicode normalization in tree conversion \|    9      9  0.0s
  | └ Collect (Basic: ✓ 1373, REPL 40/40: ✓ 1810) => Execute ✓ 2398
  | Precompilation complete. Summary:
  | Total ───────  92.991485 seconds

```

c42f commented:
> new stuff added to the precompile step is outputting in a way that leaks

Thanks! It runs a subset of the JuliaSyntax tests as part of precompile.

> If the packages are broken, my experience from the pkgimages work suggests it's better to fix known issues before merging

Yes, there's a few remaining cases which will break a package or two.  One that comes to mind is the mis-parsing of `@.a+b` (broken when there's no space between the `.` and the `a` for obscure reasons). But note that the new parser is still off by default in the current PR (gated on an env variable)

c42f commented:
So one tricky thing in this PR is the desired behavior for errors from `Meta.parse()`.

Before this PR, we have that
* `Meta.parse(str, raise=false)` returns an `Expr(:error, error_msg_string)`
* `eval(Meta.parse(str, raise=false))` throws `ErrorException`.
* `Meta.parse(str)` throws `Meta.ParseError`
* The REPL reports syntax errors by using the printing of `ErrorException` which comes from `eval(Base.parse_input_line(str))`
* The `Meta.ParseError` API is simple: it just has a `msg` string

After this PR:
* `Meta.parse(str, raise=false)` returns an `Expr(:error, JuliaSyntax.ParseError(...))`
* `eval(Meta.parse(str, raise=false))` throws `JuliaSyntax.ParseError(...)`.
* `Meta.parse(str)` throws `JuliaSyntax.ParseError(...)`, except for the line at https://github.com/JuliaLang/julia/blob/578c4320d1321024142960feabf7199dece02ca2/base/meta.jl#L272
* The REPL reports syntax errors by using the printing of `JuliaSyntax.ParseError` which comes from `eval(Base.parse_input_line(str))`
* The `JuliaSyntax.ParseError` API is effectively just `showerror()` - it's got an internal representation but nobody should be relying on that to be stable.

Overall this doesn't seem terrible for an interim situation (where the new behavior is experimental and gated behind an ENV variable). But it is quite awkward that `Meta.ParseError` is no longer the relevant type.

One option for compatibility could be to add a field to `Meta.ParseError` so we don't change the type which is thrown. For example

```
struct ParseError
     msg::String
     detail::Any  # when not `nothing`, use this in `showerror` rather than msg
end
```

c42f commented:
Sketch of how we might get parser diagnostics into the runtime is over here:

https://github.com/JuliaLang/julia/issues/36547#issuecomment-1454326641

oscardssmith commented:
@StefanKarpinski why did you close this?

KristofferC commented:
How does the concrete "end game" for shipping this to users look like? We clearly don't want to keep around both parsers so there needs to be a bootstrapping story.

c42f commented:
There's several possibilities for bootstrapping but it's a whole pile of work I don't want to tackle in this PR (also, removing the flisp parser would likely save only a tiny fraction of sysimage size I suspect).

The ideal is probably to have separate compilation, but IIUC there's no concrete timeline for that. Failing that, I've considered parsing the parser and dumping it out in a textural form which can be digested by the C code. Perhaps just s-expressions that the flisp runtime can already read, as we have C code for converting that into `Expr` and `eval`ing it. But the catch here is we'd need to include a chunk of Base in the dump.  Just as a chunk of Base is already duplicated in Core.Compiler. So hmm, maybe that's not the best idea! If we had a way to put all that in a temporary module and not save it into the sysimage after bootstrap that could work well. I don't know if this is possible though.

oscardssmith commented:
@nanosoldier `runtests(ALL, vs = "%self", configuration = (buildflags=["DEPS_GIT=JuliaSyntax"], environment=["JULIA_USE_NEW_PARSER=true"]))`

nanosoldier commented:
Your job submission was not accepted. Consult the server logs for more details (cc @maleadt).

oscardssmith commented:
I think this is ready to merge. Hopefully for 1.11 we can default it on.

KristofferC commented:
Would be good with input from @c42f heh. What's the reason it cannot default to on? Are there known bugs?

maleadt commented:
`environment` was renamed to `env`:

@nanosoldier `runtests(ALL, vs = "%self", configuration = (buildflags=["DEPS_GIT=JuliaSyntax"], env=["JULIA_USE_NEW_PARSER=true"]))`

c42f commented:
I've got some updates to push here today. But generally yes this is so so close and we should merge some version of it for 1.10.

I also think we should default it to on and see what happens. As of yesterday there were finally no known parser bugs upstream in JuliaSyntax relative to the reference parser! There's a handful of known incompatibilities for some obscure and unexpected behaviors of the reference parser which may equally well be bugs in the reference parser.

And as of recent updates it's around 10x faster too :-)

nanosoldier commented:
[Your job](https://github.com/JuliaLang/julia/pull/46372#issuecomment-1559917840) failed. Consult the server logs for more details (cc @maleadt).

maleadt commented:
`make` failed:

```
cp: cannot create regular file '/source/julia-ba69f249f4/share/julia/base/JuliaSyntax/.git/objects/pack/pack-842ec527841ee3fd1b06e17e493fbfcd0d99a01d.pack': Permission denied
cp: cannot create regular file '/source/julia-ba69f249f4/share/julia/base/JuliaSyntax/.git/objects/pack/pack-842ec527841ee3fd1b06e17e493fbfcd0d99a01d.idx': Permission denied
```

Let's try again now that some commits have been added:

@nanosoldier `runtests(ALL, vs = "%self", configuration = (buildflags=["DEPS_GIT=JuliaSyntax"], env=["JULIA_USE_NEW_PARSER=true"]))`

nanosoldier commented:
[Your job](https://github.com/JuliaLang/julia/pull/46372#issuecomment-1560489283) failed. Consult the server logs for more details (cc @maleadt).

KristofferC commented:
What's `"DEPS_GIT=JuliaSyntax"` supposed to do here?

maleadt commented:
> What's `"DEPS_GIT=JuliaSyntax"` supposed to do here?

I guess we can use the released version now. JuliaSyntax.jl is also the default as per the latest commits, so we do a simplified invocation:

@nanosoldier `runtests()`

c42f commented:
The doc build has failed due to the distinction between `JuliaSyntax.ParseError` vs `Meta.ParseError` mentioned in https://github.com/JuliaLang/julia/pull/46372#issuecomment-1423703500 and an assumption made about that in the `Markdown` stdlib.  So it's clear that needs to be fixed.  I'll try the fix with the `detail` field mentioned in that comment.

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/46372#issuecomment-1560530320) has completed - possible new issues were detected.
A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/c53e8d1_vs_c470dc3/report.html).

c42f committed:
Enable JuliaSyntax.jl as the defult parser

* Vendor JuliaSyntax into Base via deps directory
* Install JuliaSyntax as the Julia parser unless the environment
  variable JULIA_USE_NEW_PARSER=0 is set.
* Add a function to set the Core._parse binding.
  Required because we'd like to set the binding during Base.__init__.
  This can be done with `Core.eval` but that doesn't work well in
  incremental compilation mode.

Also accommodate JuliaSyntax within tests:

* When JuliaSyntax is enabled, ignore error messages in parser tests
  which are tested separately upstream - error messages are inherently
  expressed a bit differently when they go alongside full source
  location info.
* Accommodate a small number of incompatibilities where in JuliaSyntax
  - `import .Mod.x as (a.b)` is a syntax not lowering error
  - `f(2x for x=1:10, y` is `Expr(:incomplete)` not `Expr(:error)`
  - `incomplete_tag` is more precise for `:block` vs `:other`
  - `global const` without an assignment is a syntax error, in keeping
    with plain `const` without assignment being a syntax error (not
    lowering error).
* Adjust a few tests to be more precise about testing lowering vs the
  parser.
* Make Meta.parse doctest compatible with JuliaSyntax errors

c42f committed:
Make incomplete_tag extensible

This allows `incomplete_tag` to work when Expr(:incomplete) holds a
Meta.ParseError as its child rather than a plain string, as it will when
JuliaSyntax is enabled.

c42f committed:
Add `Meta.ParseError` detail field

Here we add a `detail` field to `Meta.ParseError`, but retain the
`msg::String` field for compatibility. `showerror(::ParseError)` defers
to the `detail` field if it's present.

This allows us to still throw `Meta.ParseError` from `Meta.parse` for
compatibility, but allow more expressivity when `detail` is set to an
exception type like `JuliaSyntax.ParseError`.

c42f committed:
Only wrap Strings in Meta.ParseError

String errors come from the flisp parser as Expr(:error). But other than
that, allow the parser library to choose its own error type.

c42f committed:
Show top level location for any `exc` in `eval(Expr(:error, exc))`

Previously this only worked when `exc` was a `String`.

c42f committed:
Fix string escaping in REPL completion of paths

REPL completion of paths within strings need to be escaped according to
the usual escaping rules, and delimited by the starting " rather than
whitespace.

This differs from completion of paths within cmd backticks which need to
be escaped according to shell escaping rules.

Separate these cases and fix string escaping.

This was found because JuliaSyntax emits an Expr(:error) rather than
Expr(:incomplete) for paths inside strings with invalid escape sequences
before whitespace.

c42f committed:
Pass Int rather than UInt as lengths to core parser hook

This is more consistent with the way we're likely call it from the Julia
side via Meta.parse().

c42f commented:
Turning on the parser by default surfaced a bunch of minor issues which are fixed now, except for a subtle one related to coverage.

IIUC the existing coverage tests are currently broken because
* `test/testhelpers/coverage_file.jl` uses `rand()`
* The test is effectively using `--code-coverage=user`, and stdlibs are considered "user" code
* But no coverage within `Random` is recorded with the reference parser

However, coverage within `Random` is now recorded with JuliaSyntax as the parser.  So it's possible the new parser is fixing something here, probably to do with line numbers.  But I haven't discovered what yet.

oscardssmith commented:
what does this mean wrt merging? Should we change the test?

c42f commented:
> what does this mean wrt merging? Should we change the test?

It wasn't what I thought. What's actually going on is much less interesting: The reference parser generates a trailing line number in the `Expr(:block)` of the `for` loop in the coverage test. A line number which is the location of the `end`:

```
julia> Meta.parse("""
       for i = 1:3
           push!(r, i)
       end""")
:(for i = 1:3
      #= none:2 =#
      push!(r, i)
      #= none:3 =#
  end)
```

I don't fully understand why this generates different coverage data, but I've just changed JuliaSyntax to also generate this line number node for compatibility.

c42f commented:
@nanosoldier `runtests()`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/46372#issuecomment-1574777459) has completed - possible new issues were detected.
A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/9fe8d77_vs_aea56a9/report.html).

c42f commented:
Pkg test failures here are due to https://github.com/JuliaLang/Pkg.jl/pull/3507. If that's merged and Pkg bumped these should go away.

IanButterworth commented:
Pkg is now bumped with that fix

c42f commented:
Thanks Ian :)

Oh my gosh look at those green ticks in CI.  It's been a slog.

Next up, looking closer at PkgEval

c42f commented:
I've triaged the PkgEval failures. Found a couple more obvious bugs to fix and a bunch of other issues where the right fix is less obvious.

### Bugs in JuliaSyntax

* [x] `DataPipes` had a problem with the weird `global (x,y)` syntax. Fixed by https://github.com/JuliaLang/JuliaSyntax.jl/pull/303

### Bugs in JuliaSyntax integration with Base

* [ ] `StrBase`, `EditorsRepo`, `StippleUI` try to call `Meta.parse()` with a custom string type, but this fails because the custom string type enters the parser hook which runs at a fixed world age. `GenieFramework` fails transitivly by depending on one of these.

### Minor parser changes which break package compilation

* [x] `QuadraticFormsMGHyp` accidentally used the syntax `(a; b,)` with trailing comma for a block rather than a tuple with parameters. It's the only package with this syntax. Fixed upstream in https://github.com/s-broda/QuadraticFormsMGHyp.jl/pull/8.
* [ ] `NiLang` and `NiLangCore` use some very weird syntax in a few small places for "function names" that the reference parser seems to support and then process this with a macro to remove the weirdness. See https://github.com/GiggleLiu/NiLang.jl/issues/83. I don't know whether to try to be compatible with this syntax which could have no meaning in a normal Julia program.
* [ ] `TreeverseAlgorithm`, `SparseADRules` are broken transitively by depending on `NiLang`

### Minor parser changes which break package tests

* [ ] ExprTools and Expronicon go to great lengths to handle the various anonymous and short-form function syntaxes which the reference parser is not very consistent about (and which are currently patched up during lowering). JuliaSyntax parses these forms more consistently but that breaks some of these package tests in the process.
* [ ] `Fluxsperimental` expects `kw` arguments with `do` and macro calls as in `@x(a=1) do y end`. But JuliaSyntax currently chooses to be consistent with normal macro calls like `@x(a=1)` which use `=` rather than `kw`.
* [x] `SLEEFPirates` and `SLEEF` have the same weird syntax in their tests where a string is juxtaposed with `for` as in `"name"for x in xs end`.  `SLEEFPirates` is fixed upstream but `SLEEF` seems abandoned.

### Changes to the way parse errors are represented and printed

* [ ] Revise, SoftGlobalScope both expect `eval(Meta.parse(invalid_code, raise=false))` to throw `ErrorException` but we now throw `Base.ParseError` with detailed error location info.
* [ ] Revise also fails because `LoadError` sometimes gets a different line number.
* [ ] Weave expects `showerror(io, try Meta.parse(invalid_code) catch exc ; exc end)` to not include the file name but it does

### Differences in LineNumberNodes

* [ ] CoverageTools tests fail, likely due to extra or different line number nodes.

### Cases requiring more investigation

* [ ] `ContextManagers` Everything explodes by I don't know why yet ??
* [ ] `Fable` is broken because `CUDA` doesn't precompile
* [ ] `FindDefinition` fails, speculatively due to line numbers

### Likely spurious failures

* [ ] Various tests exceeded the time limit
* [ ] `IMFData` and `UltraDark` hung during testing ??
* [x] VCFTools seems to fail tests non deterministically and the test is related to data file parsing
* [x] LuaCall test related to GC seems to work on dev version of the package
* [x] NumericalAlgorithms tests pass locally, presumably are non deterministic

longemen3000 commented:
Hmmm, if I remember correctly, tkf hasn't been available for a really long time, so that PR to ContextManagers.jl is unlikely to be merged in the foreseeable future 😕

JeffBezanson commented on test/backtrace.jl:
```diff
@@ -221,7 +228,6 @@ let trace = try
     end
     @test trace[1].func === Symbol("top-level scope")
     @test trace[1].file === :a_filename
-    @test trace[1].line == 2
```
I think this can be brought back; seems to work.

c42f replied:
Will fix

c42f commented:
> I think we can finally merge this?

Yes! This is going to make it by feature freeze :relieved: :tada: 

Let me push another couple of fixes first though, before I merge :-)

The `FindDefinitions` failure was a bug in JuliaSyntax which has been subtle to fix but I'm nearly done with that.

c42f commented:
Any idea what's up with the i686 windows build?

We seem to have an LLVM out of memory error.  But I think a previous build on this same branch succeeded, then failed, then succeeded again with this same error on the 8686 build.

So it might be build environment related?

```
Sysimage built. Summary:
Base ────────  67.772410 seconds 53.4589%
Stdlibs ─────  59.000268 seconds 46.5394%
Total ─────── 126.774893 seconds
make[1]: Leaving directory '/c/workdir'
make[1]: Entering directory '/c/workdir'
 cd /c/workdir/base && if ! JULIA_BINDIR=`cygpath -w /c/workdir/usr/bin` WINEPATH="`cygpath -w /c/workdir/usr/bin`;$WINEPATH" JULIA_LOAD_PATH='@stdlib' JULIA_PROJECT= JULIA_DEPOT_PATH=':' JULIA_NUM_THREADS=1  /c/workdir/usr/bin/julia.exe -O3 -C "pentium4;sandybridge,-xsaveopt,clone_all" --output-o `cygpath -w /c/workdir/usr/lib/julia/sys-o.a`.tmp  --startup-file=no --warn-overwrite=yes --sysimage `cygpath -w /c/workdir/usr/lib/julia/sys.ji` `cygpath -w /c/workdir/contrib/generate_precompile.jl` 1; then echo '*** This error is usually fixed by running `make clean`. If the error persists, try `make cleanall`. ***'; false; fi 
Collecting and executing precompile statements
└ Collect (Basic: ✓ 819, REPL LLVM ERROR: out of memory
Allocation failed

[7388] signal (22): SIGABRT
in expression starting at none:0
crt_sig_handler at C:/workdir/src\signals-win.c:100
raise at C:\Windows\System32\msvcrt.dll (unknown line)
abort at C:\Windows\System32\msvcrt.dll (unknown line)
.text$_ZN4llvm22report_bad_alloc_errorEPKcb at C:\workdir\usr\bin\libLLVM-15jl.dll (unknown line)
Allocations: 86200899 (Pool: 86152611; Big: 48288); GC: 250
*** This error is usually fixed by running `make clean`. If the error persists, try `make cleanall`. ***
make[1]: *** [sysimage.mk:93: /c/workdir/usr/lib/julia/sys-o.a] Error 1
make[1]: Leaving directory '/c/workdir'
🚨 Error: The command exited with status 2
^^^ +++
^^^ +++
user command error: The plugin docker command hook exited with status 2
~~~ Running plugin docker pre-exit hook
C:\buildkite-agent\plugins\github-com-buildkite-plugins-docker-buildkite-plugin-v3-13-0\hooks\pre-exit
cd C:\buildkite-agent\builds\win2k22-amdci6-7\julialang\julia-master\c\buildkite-agent\builds\win2k22-amdci6-7\julialang\julia-master
```

oscardssmith committed:
Merge branch 'master' into cjf/juliasyntax-stdlib

giordano commented:
@gbaraldi said that error has been seen in other PRs, so it doesn't seem to be specific to this one.

gbaraldi commented:
Well, I've seen it in https://github.com/JuliaLang/julia/pull/50144, which is a PR that does mess with the GC, but I do think we near the limit when compiling the sysimg at 32 bit.

giordano commented:
#50205 it's broken also on master.

c42f commented:
@giordano thanks, what a relief. I also tried to repro the LLVM out of memory on the i686 windows build locally with cygwin, and could not.

c42f commented:
I think this is good to go. I'll do one last nanosoldier run and have a look through those before merging.

@nanosoldier `runtests()`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/46372#issuecomment-1596008232) has completed - possible new issues were detected.
A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/fde8692_vs_65523e4/report.html).

oscardssmith committed:
Merge branch 'master' into cjf/juliasyntax-stdlib

oscardssmith commented:
Thank you so much @c42f for all the work to make this a reality!

c42f commented:
Ah. I must admit I was really looking forward to pushing the merge button on this myself.  (Not if it meant missing feature freeze, but that's tomorrow morning my time IIUC?)

I was just preparing a final minor bugfix based on the latest nanosolider run. I'll make a separate PR for that.

aviatesk commented:
I think there may be some elements, such as the method of merging, that would be better handled by @c42f . If you feel it's preferrable, we wouldn't mind if the PR were to be temporarily reverted and reland it quickly with minor fixes, what are your thoughts? When it comes to major PRs like this one, it often seems best if the person who created the PR decides matters such as whether to squash the commits. Regarding the feature freeze, it's set to be extended until the 27th of June, so there's no need to get hurry just yet:)

c42f commented:
At this point I'd rather not revert this - that will just cause extra churn. I think it'll be fine just to put a separate PR in for any minor fixes because this was very, very close and the remaining bug I know about doesn't seem very disruptive.

Luckily I'd already rearranged the commits into logical order and done a bunch of squashing so that side of it was in pretty good shape.
